src/modules/ekf2/EKF/ev_height_control.cpp

```diff
@@ -38,94 +38,81 @@
+	// increase minimum variance if GPS active
+	if (_control_status.flags.gps_hgt) {
+		measurement_var = math::max(measurement_var, sq(_params.gps_pos_noise));
+	}
 
-		if (_control_status.flags.ev_hgt) {
-			aid_src.fusion_enabled = true;
+	const bool measurement_valid = PX4_ISFINITE(measurement) && PX4_ISFINITE(measurement_var);
 
-			if (continuing_conditions_passing) {
+	updateVerticalPositionAidSrcStatus(ev_sample.time_us,
+					   measurement - bias_est.getBias(),
+					   measurement_var + bias_est.getBiasVar(),
+					   math::max(_params.ev_pos_innov_gate, 1.f),
+					   aid_src);
 
-				fuseVerticalPosition(aid_src);
+	// update the bias estimator before updating the main filter but after
+	// using its current state to compute the vertical position innovation
+	if (measurement_valid && quality_sufficient) {
+		bias_est.setMaxStateNoise(sqrtf(measurement_var));
+		bias_est.setProcessNoiseSpectralDensity(_params.ev_hgt_bias_nsd);
+		bias_est.fuseBias(measurement - _state.pos(2), measurement_var + P(9, 9));
+	}
 
-				const bool is_fusion_failing = isTimedOut(aid_src.time_last_fuse, _params.hgt_fusion_timeout_max);
-
-				if (isHeightResetRequired()) {
-					// All height sources are failing
-					ECL_WARN("%s height fusion reset required, all height sources failing", HGT_SRC_NAME);
-
-					_information_events.flags.reset_hgt_to_ev = true;
-					resetVerticalPositionTo(measurement - bias_est.getBias(), measurement_var);
-					bias_est.setBias(-_state.pos(2) + measurement);
-
-					// reset vertical velocity
-					if (ev_sample.vel.isAllFinite() && (_params.ev_ctrl & static_cast<int32_t>(EvCtrl::VEL))) {
-
-						// correct velocity for offset relative to IMU
-						const Vector3f pos_offset_body = _params.ev_pos_body - _params.imu_pos_body;
-						const Vector3f vel_offset_body = _ang_rate_delayed_raw % pos_offset_body;
-						const Vector3f vel_offset_earth = _R_to_earth * vel_offset_body;
-
-						switch (ev_sample.vel_frame) {
-						case VelocityFrame::LOCAL_FRAME_NED:
-						case VelocityFrame::LOCAL_FRAME_FRD: {
-								const Vector3f reset_vel = ev_sample.vel - vel_offset_earth;
-								resetVerticalVelocityTo(reset_vel(2), math::max(ev_sample.velocity_var(2), sq(_params.ev_vel_noise)));
-							}
-							break;
-
-						case VelocityFrame::BODY_FRAME_FRD: {
-								const Vector3f reset_vel = _R_to_earth * (ev_sample.vel - vel_offset_body);
-								const Matrix3f reset_vel_cov = _R_to_earth * matrix::diag(ev_sample.velocity_var) * _R_to_earth.transpose();
-								resetVerticalVelocityTo(reset_vel(2), math::max(reset_vel_cov(2, 2), sq(_params.ev_vel_noise)));
-							}
-							break;
-						}
+	const bool continuing_conditions_passing = ((_params.ev_ctrl & static_cast<int32_t>(EvCtrl::VPOS))
+			|| (_params.height_sensor_ref == HeightSensor::EV))
+			&& measurement_valid;
 
-					} else {
-						resetVerticalVelocityToZero();
-					}
+	starting_conditions_passing &= continuing_conditions_passing;
```

Comment:
```suggestion
	const bool starting_conditions_passing = mandatory_conditions_passing && continuing_conditions_passing;
```
Maybe pass them as "mandatory_conditions_passing" or "general EV conditions"

Reply:
Updated to common_starting_conditions_passing.